probe-rs/src/core/registers.rs

```diff
@@ -293,151 +351,220 @@ impl<T> RegisterValueResultExt<T> for Result<T, Infallible> {
+                ))
+            })
     }
 
     /// The main stack pointer.
-    pub fn msp(&self) -> Option<&RegisterDescription> {
-        self.msp
+    pub fn msp(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::MainStackPointer))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No main stack pointer found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// The process stack pointer.
-    pub fn psp(&self) -> Option<&RegisterDescription> {
-        self.psp
+    pub fn psp(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::ProcessStackPointer))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No process stack pointer found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// The processor status register.
-    pub fn psr(&self) -> Option<&RegisterDescription> {
-        self.psr
+    pub fn psr(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::ProcessorStatus))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No processor status register found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// Other architecture specific registers
-    pub fn other(&self) -> impl Iterator<Item = &RegisterDescription> {
-        self.other.iter()
+    pub fn other(&self) -> impl Iterator<Item = &CoreRegister> {
+        self.0
+            .iter()
+            .filter(|r| matches!(r.role, Some(RegisterRole::Other(_))))
     }
 
     /// Find an architecture specific register by name
-    pub fn other_by_name(&self, name: &str) -> Option<&RegisterDescription> {
-        self.other.iter().find(|r| r.name == name)
+    pub fn other_by_name(&self, name: &str) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| matches!(r.role, Some(RegisterRole::Other(other_name)) if other_name == name))
+            .ok_or_else(|| {
+                Error::GenericCoreError(format!(
+                    "Argument register {name:?} not found. Please report this as a bug."
+                ))
+            })
     }
 
     /// The fpu status register.
-    pub fn fpscr(&self) -> Option<&RegisterDescription> {
-        self.fp_status
+    pub fn fpsr(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::FloatingPointStatus))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No FPU status register found. Please report this as a bug.".to_string(),
```

Comment:
This is only a bug if the processor has an FPU, so this should return `None` if no register is found.

Reply:
You are absolutely right. I was a bit too focussed on harmonizing behaviour, `Option<>` makes more sense for these. Consider it done :)